### DISCUSSION TITLE
Decompiling AST

### DISCUSSION THREAD

**Discussion by eadgbear** [rhaiscript/rhai#698] [General]
Not exactly the best title, I know.

Is it possible to take an AST and turn it back into Source? Not the exact formatting or anything like that, just turn it back into the source code. I have an experiment where I'm modifying the AST and want to show it back to the user.

**Comment by schungx** [comment]
Not easily as Rhai's AST does not contain all the contextual information required to reconstruct a source text.

Mainly this has to do with the `switch` statement but things like custom syntax etc. are also troublesome.

That's why the compression feature uses token manipulation instead of writing out and AST as text.

**Comment by schungx** [comment]
If you're willing to be only about 80% of the way there, then it is not difficult to write. You just have to be careful about parentheses.

**Comment by schungx** [comment]
Closing this for now. Feel free to reopen if necessary.
